### PR TITLE
Remove deprecated FailureReason enum values

### PR DESCRIPTION
**Pull request by pmossman**
### Overview

The current FailureReason schema has a few issues:

1. We decided to replace `replicationWorker` origin with `replication`, but when I made this change initially, I neglected to consider records that had already been written to databases with `replicationWorker` set as the value. When we added failure origins to the UI, any workspace that had a `replicationWorker` enum value set would crash due to deserialization failure. In order to get us back online quickly, I re-added `replicationWorker` to the enum.

2. We decided that having an 'unknown' value for a nullable field was confusing since 'unknown' and 'null' mean the same thing. We don't actually set 'unknown' anywhere in code, so this PR just removes it from the enum to avoid confusion.

3. There is a lurking snake_case vs camelCase discrepancy in the `failureType` enum that could cause a deserialization failure in the future. The API schema uses snake_case while the json schema uses camelCase. Right now we don't have any code that depends on this enum but this will likely change in the near future. This PR updates the enum values to snake_case so that we're consistent between what we store in the DB and what the API expects.

Basically, there were some pitfalls around working with json schemas, jsonb columns, and serialization that I fell into during this project, and now that I've learned what those pitfalls are and how to avoid them, I'm writing this PR to clean up the mess and get us back into a solid state.

### Implementation Details

- The migration queries the `attempts` table to find all records that contain a failure summary that contains an enum value that needs correcting.
- The migration class defines its own version of the AttemptFailureSummary and FailureReason class that are essentially copies of the real thing, except for the fact that they support both the old and new versions of the enum values. This is important because we need this migration to be able to deserialize records whenever it is run, no matter what version of Airbyte it's being run against. I can't use the real version of these classes as deserialization targets because we're removing the deprecated enum values from them.
- The migration test should cover every bad enum value and the update from 'unknown' to null.

### CODE REVIEW

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+      "failureType",
+      "internalMessage",
+      "externalMessage",
+      "metadata",
+      "stacktrace",
+      "retryable",
+      "timestamp"
+  })
+  static
+  class FailureReasonForMigration {
```
Comment (lmossman) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java:145]: The formatting is a little weird here - why are `static` and `class` on different lines?

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+  }
+
+
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  @JsonPropertyOrder({
+      "failures",
+      "partialSuccess"
+  })
+  static
+  class AttemptFailureSummaryForMigration implements Serializable {
```
Comment (lmossman) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java:461]: These classes will still work even if new enum values or new json properties are added to the db, right? I'm guessing any new values would be ignored?

Or maybe that shouldn't matter because this migration will not be ran after changes are added in future versions?

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_5_001__Add_failureSummary_col_to_Attempts.java [hunk 3]
```diff
@@ -21,6 +22,11 @@ public void migrate(final Context context) throws Exception {
     LOGGER.info("Running migration: {}", this.getClass().getSimpleName());
 
     final DSLContext ctx = DSL.using(context.getConnection());
+    migrate(ctx);
+  }
+
+  @VisibleForTesting
+  public static void migrate(final DSLContext ctx) {
```
Comment (lmossman) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_5_001__Add_failureSummary_col_to_Attempts.java:29]: What is the purpose of this change?

Comment (lmossman) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_5_001__Add_failureSummary_col_to_Attempts.java:29]: Ah just saw that this is used in the test below, disregard

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java (hunk 2 above)
Comment (pmossman) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java:461]: Good call, I should write a test case for what happens if the DB contains a value that isn't known by this class. Even if it _shouldn't_ matter according to changes being added in future versions, it should be possible to make the migration resilient to it.

**Review by lmossman — APPROVED** [review]
A couple small comments but nothing blocking, LGTM

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_5_001__Add_failureSummary_col_to_Attempts.java (hunk 3 above)
Comment (pmossman) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_5_001__Add_failureSummary_col_to_Attempts.java:29]: I think this is a better pattern for making migrations callable from unit tests (I took it from https://github.com/airbytehq/airbyte/blob/master/airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_32_8_001__AirbyteConfigDatabaseDenormalization.java#L62-L75). 

That said, it does seem strange to modify migrations that we've already run, so maybe I'll revert it here

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java (hunk 1 above)
Comment (pmossman) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_35_40_001__MigrateFailureReasonEnumValues.java:145]: Oh weird, maybe copy/paste and autoformatting did something wonky

**Comment by pmossman** [comment]
@lmossman I ended up tweaking the deserialization classes to use `String` fields instead of enums for `failureOrigin` and `failureType`. This way, any database record should be able to deserialize into those classes when this migration is run, and we can still update the old values to the new values like before. I also added a unit test case that ensures we tolerate some random string value in the database.

@cgardens I definitely would appreciate you sanity checking this before I merge, I've been trying to wrap my head around edge cases where OSS users jump many versions forward and whether or not this migration could break if we end up changing our FailureReason json schema in the future. 

I _think_ this migration should be robust to that case because the only records we'll ever attempt to deserialize must have been added in a previous Airbyte version, but I wouldn't be surprised if there's an edge case I'm not considering.

**Review by cgardens — APPROVED** [review]
looks good to me. I agree with your assessment. 

Code will only run with the new enum values _after_ the migration you wrote run. The contract is migration runs (via the bootloader) before new version before any new code executes. So as long as there is nothing in the bootloader itself that uses these values before the migration is run, then this approach should work. 

I didn't go through the migration line-by-line, but I think copying the old values / code into the migration code makes sense to me as well.
